### PR TITLE
⚠️  clusterctl completion zsh: Use native zsh completion

### LINKED ISSUE #4090: clusterctl completion: Use native zsh completion instead
**User Story**

As a developer I would like to use native zsh completion instead of the bash-compatibility one for making it easier to maintain.

**Detailed Description**

The current zsh completion of clusterctl is achieved by replacing some of the bash completions. Only a limited number of developers can maintain it because it is difficult to understand how it works.

- https://github.com/kubernetes-sigs/cluster-api/blob/v0.3.12/cmd/clusterctl/cmd/completion.go#L122-L219

Since cobra has supported native zsh completion since v1.1, we can replace it with this and it will be much easier to maintain. In addition, we will be ready to provide complements to other shells such as fish in the future.

**Anything else you would like to add:**

The native zsh completion provided by cobra does not support the use of completion in the following method:

```bash
source <(clusterctl completion zsh)
```

Instead, we should do the following. clusterctl completion's help message suggests the above method is available and will require the user to make changes.

```bash
$ clusterctl completion zsh > "${fpath[1]}/_clusterctl"
```

- https://github.com/spf13/cobra/blob/master/shell_completions.md#zsh-completions

/kind feature
/area clusterctl

### PR DESCRIPTION
**Pull request by superbrothers**
**What this PR does / why we need it**:

ref/ #4090 

Due to this change, we will not able to load the zsh completion code
with the following method:
```
source <(clusterctl completion zsh)
```

Instead, we execute the following command once:
```
clusterctl completion zsh >"${fpath[1]}/_clusterctl"
```
Then we need to start a new shell for this setup to take effect.

---

In addition, github.com/spf13/cobra is updated to v1.1.1 in order to use
native zsh completion feature.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Closes #4090

### CODE REVIEW

**Comment by prankul88** [comment]
/assign

Thanks for this PR. I can review this to check functionality.

File: cmd/clusterctl/cmd/completion.go [hunk 1]
```diff
@@ -73,20 +62,29 @@ var (
...
 
-		# Load the clusterctl completion code for zsh[1] into the current shell
-		source <(clusterctl completion zsh)`)
+		Zsh:
+		# If shell completion is not already enabled in your environment you will need
+		# to enable it.  You can execute the following once:
+		echo "autoload -U compinit; compinit" >> ~/.zshrc
+
+		# To load completions for each session, execute once:
+		clusterctl completion zsh > "${fpath[1]}/_clusterctl"
```
Comment (prankul88) [inline cmd/clusterctl/cmd/completion.go:71]: I am getting a permission error when running this command, 
>zsh: permission denied: /usr/local/share/zsh/site-functions/_clusterctl

Not sure if only it is system-specific.

File: cmd/clusterctl/cmd/completion.go [hunk 2]
```diff
@@ -73,20 +62,29 @@ var (
...
+
+		# You will need to start a new shell for this setup to take effect.`)
```
[the remark this answers is not in the document]
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go:73]: I've updated the capi book at https://github.com/superbrothers/cluster-api/commit/2a2c89db5a825aaacaca79e3c46a790a19cf54a1.

File: cmd/clusterctl/cmd/completion.go (hunk 1 above)
Comment (superbrothers) [reply cmd/clusterctl/cmd/completion.go:71]: It depends on the system.

File: go.sum.orig [hunk 3]
```diff
@@ -0,0 +1,967 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (wfernandes) [inline go.sum.orig:1]: Here too. I'm guessing we don't need this file as well. 🙂

**Review by wfernandes — CHANGES_REQUESTED** [review]
@superbrothers Thanks for this update! I haven't been able to test the zsh completion but it mostly lgtm just some nits and clarifications regarding some files.

File: go.mod.orig [hunk 4]
```diff
@@ -0,0 +1,48 @@
+module sigs.k8s.io/cluster-api
```
[the remark this answers is not in the document]
Comment (superbrothers) [reply go.mod.orig:1]: I made a mistake with `git rebase` operation😅

[the remark this answers is not in the document]
Comment (superbrothers) [reply go.mod.orig]: I've fixed it and squashed the commits.

**Comment by CecileRobertMichon** [comment]
```
➜  cluster-api git:(pr/4113) bin/clusterctl 

alpha       -- Commands for features in alpha.
completion  -- Output shell completion code for the specified shell (bash or
config      -- Display provider configuration and templates to create worklo
delete      -- Delete one or more providers from the management cluster.
describe    -- Describe workload clusters.
generate    -- Generate yaml using clusterctl yaml processor.
get         -- Get info from a management or a workload cluster
help        -- Help about any command
init        -- Initialize a management cluster.
move        -- Move Cluster API objects and all dependencies between managem
upgrade     -- Upgrade core and provider components in a management cluster.
version     -- Print clusterctl version.

```

Works for me!

/lgtm

**Comment by CecileRobertMichon** [comment]
Just one question: why is the change marked as breaking ⚠️ ?

**Comment by wfernandes** [comment]
@CecileRobertMichon It seems that the loading of the zsh logic (from the changes in the docs) has changed a bit.

**Comment by CecileRobertMichon** [comment]
> @CecileRobertMichon It seems that the loading of the zsh logic (from the changes in the docs) has changed a bit.

Ok, just wanted to make sure there wasn't another breaking change I had missed

/approve

**Comment by marckhouzam** [comment]
> Just one question: why is the change marked as breaking ⚠️ ?

> Due to this change, we will not able to load the zsh completion code with the following method:
> ```
> source <(clusterctl completion zsh)
> ```

Zsh recommends putting the completion file on `fpath` instead of sourcing it, which is why Cobra does not support sourcing the completion file.  This is better for performance as it enables zsh to do some sort of caching of the script.

It is however possible to support sourcing if you feel it is better for your users. To do that you have to add the line `compdef _clusterctl clusterctl` to the completion script (at the end of the script is safe).  This is what the Helm project has chosen to do to avoid a "breaking" change: https://github.com/helm/helm/blob/037a43c98bc6331ffe7027d89e57fb1334cc1189/cmd/helm/completion.go#L168

**Comment by wfernandes** [comment]
@marckhouzam We have this in our documentation https://master.cluster-api.sigs.k8s.io/clusterctl/commands/completion.html#zsh which seems like it adds it to the fpath.

Also the output of `clusterctl completion zsh` has `compdef _clusterctl clusterctl` as the first line.

So I guess this isn't a breaking change. Yay!

**Comment by marckhouzam** [comment]
> @marckhouzam We have this in our documentation https://master.cluster-api.sigs.k8s.io/clusterctl/commands/completion.html#zsh which seems like it adds it to the fpath.

Well, it tells users to add it themselves 😁.  This is great and may be what you prefer to stick with.  I just wanted to mention there is a way to enable sourcing of you feel it is important.

> Also the output of `clusterctl completion zsh` has `compdef _clusterctl clusterctl` as the first line.

I believe the first line starts with a `#`, which makes the line a directive for zsh when the script is on `fpath`.  You would need a second one without the `#` for sourcing to work.

But I'm not arguing in favor, I just wanted to inform as it is something that has come up a couple of times in Cobra Issues.

**Comment by superbrothers** [comment]
> It is however possible to support sourcing if you feel it is better for your users. To do that you have to add the line compdef _clusterctl clusterctl to the completion script (at the end of the script is safe). This is what the Helm project has chosen to do to avoid a "breaking" change:

This is a nice technique. We should use this technique to avoid a "breaking change".

**Pull request merged**